### DISCUSSION TITLE
Would openEMS be suitable for small low frequency item?

### DISCUSSION THREAD

**Discussion by davidkirkby** [thliebig/openEMS-Project#105] [Q&A Usage]
I know it's not the main aim of openEMS, and I'm aware of other simpler tools, but would it be possible to compute the capacitance of a parallel plate capacitor with electrodes 100 mm diameter, 10 mm thick, at various spacings between the plates? I'd be looking for a low frequency solution, (of the order of Hz), so these dimensions are much smaller than a wavelength. I don't think it would change much if the frequency was increased to kHz, and perhaps even a MHz, but I was looking to compare to static (DC) solutions. I don't know what limitations the FDTD might put on this sort of thing. I realise it can't actually do a DC simulation.

**Comment by LubomirJagos** [comment]
other people will probably gives you more suitable answer but just out of my head, I think that DC simulation in this case is electro-static simulation, for that there are various solvers like:
  - FastCap (https://www.fastfieldsolvers.com/software.htm) + FreeCAD this need little extra work like there is no direct connection like in case for inductance in FreeCAD for FastHenry, workbench for FreeCAD (https://wiki.freecad.org/EM_Workbench/en)
  - Agros http://www.agros2d.org/
  - FEMM maybe? it's for magnetics mainly (https://www.femm.info/wiki/ElectrostaticsTutorial)
  
About openEMS for this purpose I was thinking about that but it's using FDTD what is full-wave simulation which simulates wave propagation in time and the case what you want to simulation is static in time therefore it's electrostatic case.

This is some random presentation what I found on internet https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwjBzYOvw8GAAxUHgv0HHfE5BB8QFnoECC0QAQ&url=https%3A%2F%2Fwww.researchgate.net%2Fprofile%2FStephen_Poon3%2Fpost%2FHello-every-one-If-any-one-know-about-3D-car-model-in-HFSS-or-in-MAXWELL-user-defined-or-its-design-please-tell-me-about-that%2Fattachment%2F59d6512979197b80779a9c8a%2FAS%253A506519802179589%25401497775140000%2Fdownload%2FElectromagnetic%2Bfield%2Bsimulation.pdf&usg=AOvVaw33520Xu_eXmK43kaArMnJo&opi=89978449|

But it says that depends from frequency what you are simulation there are (ordered from DC to higher frequencies) different methods:
    - electrostatic
    - quasi-static
    - full-wave

To get answer to your question, probably somehow it would be possible to put ports to some model created but to even be able create suitable model your mesh will be really dense therefore timestep would be so small that to compute it will take days :D and you will get full wave parameters which you will have to recalculate to capacitance.

**Reply by samerps** [reply]
The EM workbench for FreeCAD contains a python script "export_mesh.py" which you can use to export a mesh to FastCap.

**Comment by davidkirkby** [comment]
Thank you for the suggestions. I am aware of some of those packages, but was wondering if it's possible to do in openEMS. I actually want to learn openEMS for some RF things, but someone I knew was working at low frequencies, and I wondered if openEMS would be capable.

**Reply by VolkerMuehlhaus** [reply]
I have done such simulations (extractions of parasitic C in layout) using another FDTD solver, so it is possible in theory. FDTD is about the worst method to get the low frequency response, and I only used FDTD because I already had the model for simulation of RF properties in that exact circuit.

The reason why it **can** work is because capacitance to be extracted is rather insensitive to frequency over a wide band, and will start increasing towards SRF where the plates are no longer small w.r.t wavelength. To make it run in FDTD, you need to go up in frequency as high as possible (here 100 MHz or so) and set a very strict convergence limit and many timesteps, to minimize the residual ripple at low frequency. The FDTD solver that I used has a feature to remove that ripple, so my simulation results did make sense and converged at low frequency towards the DC value. I never tried using openEMS, but it might work.

So give it a try, carefully check if the extracted C versus frequency looks physically meaningful, and let us know your findings!

**Reply by VolkerMuehlhaus** [reply]
I have created a quick & dirty model for square 100mm*100mm capacitor using 5mm plate distance and 5mm plate thickness. Simulation takes 20 seconds and gives ~22pF from DC to 100 MHz, increasing towards higher frequencies as expected (SRF). The variation in C at very low frequency is a simulation artefact.

Theoretical result for a plate capacitor **excluding** fringing fields is 17.7pF, so the 22pF isn't completely wrong: we have strong fringing fields from thick plates in my model. 

Good luck and check results carefully!
Volker

![platecap_sideview](https://github.com/thliebig/openEMS-Project/assets/47922981/f7bb1403-a028-4f95-a83f-2fe029576385)

![platecap_3Dview](https://github.com/thliebig/openEMS-Project/assets/47922981/d181b392-fa84-40e4-ac79-328a8434a35c)

![platecap_c](https://github.com/thliebig/openEMS-Project/assets/47922981/84540f13-5ed0-47d4-83c0-21a00d42895b)

```
# Plate capacitor test, medium air epsr=1


import os
from pylab import *

from CSXCAD  import ContinuousStructure
from openEMS import openEMS
from openEMS.physical_constants import *

# preview model/mesh only?
# postprocess existing data without re-running simulation?
preview_only = False
postprocess_only = False

# get *.py model path and put simulation files in data directory below
model_path = os.path.normcase(os.path.dirname(__file__))
model_basename = os.path.basename(__file__).replace('.py','')
data_path = os.path.join(model_path, 'data')
sim_path = os.path.join(data_path,  model_basename)

print('Model path:', model_path)
print('Model basename:', model_basename)
print('Simulation data path:', sim_path, '\n')

# create data directory if it does not exist
if not os.path.exists(sim_path):
    os.makedirs(sim_path)

# change to data directory
os.chdir(sim_path)

############ simulation settings ############
 
unit = 1e-3 # specify everything in mm

Z0 = 50  # reference impedance 

fstart = 0
fstop  = 100e6
numfreq = 501  # number of frequency points (has no effect on simulation time!)

energy_limit = -50    # end criteria for residual energy
Boundaries   = ['PEC', 'PEC', 'PEC', 'PEC', 'PEC', 'PEC']  # xmin xmax ymin ymax zmin zmax

# geometry dimensions
sizex = 100
sizey = 100
platedist = 5
platethick = 5

# simulation box oversize relative to cap, each direction
box_oversize = 200

lim = exp(energy_limit/10 * log(10))
FDTD = openEMS(EndCriteria=lim)
FDTD.SetGaussExcite( (fstart+fstop)/2, (fstop-fstart)/2 )
FDTD.SetBoundaryCond( Boundaries )

wavelength_air = (3e8/unit)/fstop
max_cellsize = wavelength_air/50 # max cellsize is lambda/50 in medium air


############ Geometry setup ############
CSX = ContinuousStructure()
FDTD.SetCSX(CSX)
mesh = CSX.GetGrid()
mesh.SetDeltaUnit(unit)

# bottom plate
MyMetal = CSX.AddMetal( 'PEC' )
MyMetal.AddBox(priority=200, start=[-sizex/2, -sizey/2, 0], stop=[sizex/2, sizey/2, -platethick])

# top plate
MyMetal.AddBox(priority=200, start=[-sizex/2, -sizey/2, platedist], stop=[sizex/2, sizey/2, platedist+platethick])

# For the port, we want to limit vertical port size, so that it doesn't become large for large plate distance
# At small plate distance modelled here, we just place the vertical  port between the plates. 
# For larger plate distance, we would instead model the path mostly with vias and only a short vertical lumped port.
port1 = FDTD.AddLumpedPort(1, Z0, [-1, -1, 0],     [1, 1, platedist],     'z', excite=1.0, priority=150)


# create mesh 

# x mesh at simulation boundaries
mesh.AddLine('x', - (sizex/2 + box_oversize))
mesh.AddLine('x',   (sizex/2 + box_oversize))

mesh.AddLine('y', - (sizey/2 + box_oversize))
mesh.AddLine('y',   (sizey/2 + box_oversize))

mesh.AddLine('z', - (platedist/2 + platethick + box_oversize))
mesh.AddLine('z',   (platedist/2 + platethick + box_oversize))

# fine mesh near ports
mesh.AddLine('x', linspace(-2 , 2, 5))
mesh.AddLine('y', linspace(-2 , 2, 5))

# some more xy mesh lines inside plates
mesh.AddLine('x', linspace(-sizex/2, -2, 15))
mesh.AddLine('x', linspace(2, sizex/2, 15))

mesh.AddLine('y', linspace(-sizey/2, -2, 15))
mesh.AddLine('y', linspace(2, sizey/2, 15))

# extra mesh lines at capacitor edges, to resolve fringing fields
mesh.AddLine('x', linspace(- sizex/2 - 2*platedist , - sizex/2, 7))
mesh.AddLine('x', linspace(  sizex/2,   sizex/2 + 2*platedist, 7))

mesh.AddLine('y', linspace(- sizey/2 - 2*platedist , - sizey/2, 7))
mesh.AddLine('y', linspace(  sizey/2, sizey/2 + 2*platedist, 7))


# z mesh between plates with a few mesh lines
mesh.AddLine('z', linspace( 0, platedist, 5))
mesh.AddLine('z', linspace( -platethick, 0, 5))
mesh.AddLine('z', linspace(  platedist, platedist+platethick, 5))

mesh.SmoothMeshLines('x', max_cellsize)
mesh.SmoothMeshLines('y', max_cellsize)
mesh.SmoothMeshLines('z', max_cellsize)


#################### write model file and view in AppCSXCAD ################

# write CSX file 
CSX_file = os.path.join(sim_path, model_basename + '.xml')
CSX.Write2XML(CSX_file)

if not postprocess_only: # preview model, but only for first port excitation
    from CSXCAD import AppCSXCAD_BIN
    os.system(AppCSXCAD_BIN + ' "{}"'.format(CSX_file))

if not preview_only:  # start simulation 
    if not postprocess_only:
        FDTD.Run(sim_path, verbose=1)



###########  post-process ###########

f = np.linspace(fstart,fstop,numfreq)

if not preview_only:
    port1.CalcPort( sim_path, f, ref_impedance = Z0)
    s11 = port1.uf_ref / port1.uf_inc
    
    # calculate C from Zin
    Zin = port1.uf_tot / port1.if_tot
    omega = 2*pi*f
    C_pF=-1E12/(omega*imag(Zin))
    

    figure()
    plot(f/1e6,C_pF,'k-',linewidth=2 , label='C [pF]')
    grid()
    legend()
    # show plots
    show()

```

**Comment by biergaizi** [comment]
No, it's not suitable.

My understanding is that FDTD is inherently unsuitable for low-frequency simulation, at least in its unmodified form. The FDTD timestep must be small enough to resolve the smallest physical feature, and also small enough to resolve the highest frequency. Thus, simulating small structures at low frequencies requires an extremely huge number of timesteps. Even if it can be computed by brute force (it's usually impractical), it's numerically unstable and causes a phenomenon called "low-frequency breakdown."

Thus, if you want to obtain the raw electric and magnetic fields at low frequencies, standard FDTD is not a suitable method. That being said, if you just want to perform capacitance extraction, running the same simulation at high frequencies is a simple solution. For RF applications, DC results can also be obtained via [DC extrapolation](https://scikit-rf.readthedocs.io/en/latest/examples/networktheory/DC%20Extrapolation%20for%20Time%20Domain%20.html) (but likely inaccurate, only useful if DC is relatively unimportant).

There are some possible modifications to avoid this problem, but most standard FDTD code does not implement them.

See:

> M. Gaffar and D. Jiao, "[On the low-frequency breakdown of FDTD](https://ieeexplore.ieee.org/document/7304555)," 2015 IEEE International Symposium on Antennas and Propagation & USNC/URSI National Radio Science Meeting, Vancouver, BC, Canada, 2015.
>
> Abstract: In this work, we find the FDTD-based time-domain simulation breaks down at low frequencies in both explicit and implicit time marching schemes. This breakdown also exists in the other time-domain methods since the root cause is finite machine precision. A rigorous and fast method is developed in an unconditionally stable and also explicit FDTD method to solve the low-frequency breakdown problem. Numerical simulations of transient phenomena with arbitrarily low input frequencies have validated the proposed method.

**Reply by PinkMushroom** [reply]
It's been a while since I looked at it, but i seem to remember that the alternating direction implicit, ADI, is capable of taking care of the minimum time step problem for small geometries

https://ieeexplore.ieee.org/document/896466

The other interesting thing, is that it seems like it would be quite straightforward, to me anyway because I'm not planning on doing the work ;-) , to make FDTD an electrostatic solver, i.e. a Poisson problem solver.  One reason that might be useful is that it would then be capable of solving conducted thermal problems.

**Reply by thliebig** [reply]
I have implemented a Poisson solver a long time ago, this is actually the reason why CSXCAD (the CAD library used by openEMS) exists ;)
But it was never very efficient and I think I lost it. But it would be more efficient to implement it new instead of using openEMS as a basis...
